config.org
#+TITLE: Emacs configuration
#+AUTHOR: Janne Sauvala
#+PROPERTY: header-args:emacs-lisp :results silent :tangle init.el

* Table of contents
:PROPERTIES:
:TOC:      :include all :ignore this
:END:
:CONTENTS:
- [[#add-lexical-binding-to-generated-file][Add lexical binding to generated file]]
- [[#native-compilation][Native compilation]]
- [[#auto-update-packages][Auto update packages]]
- [[#keep-emacsd-clean][Keep .emacs.d clean]]
- [[#server-mode][Server mode]]
- [[#setup-path-from-shell][Setup path from shell]]
- [[#keyboard-bindings][Keyboard bindings]]
  - [[#mac-keyboard--finnish-layout][Mac keyboard + Finnish layout]]
  - [[#esc-cancels-all][ESC cancels all]]
  - [[#evil][Evil]]
  - [[#which-key][which-key]]
  - [[#leader-bindings-with-generalel][Leader bindings with general.el]]
  - [[#enable-keychord-bind-with-use-package][Enable keychord bind with use-package]]
  - [[#custom-keybindings][Custom keybindings]]
- [[#general-configuration][General configuration]]
  - [[#window-size][Window size]]
  - [[#dashboard][Dashboard]]
  - [[#improve-scrolling][Improve scrolling]]
  - [[#line-numbers][Line numbers]]
  - [[#theme][Theme]]
  - [[#font][Font]]
  - [[#emojis-in-buffers][Emojis in buffers]]
  - [[#doom-modeline][Doom modeline]]
  - [[#mode-diminishing][Mode diminishing]]
  - [[#recent-files][Recent files]]
  - [[#restarting-and-quitting-emacs][Restarting and quitting Emacs]]
- [[#completion][Completion]]
  - [[#vertico][Vertico]]
  - [[#corfu][Corfu]]
  - [[#save-minibuffer-history][Save minibuffer history]]
  - [[#completion-metadata-with-marginalia][Completion metadata with Marginalia]]
  - [[#improve-completions][Improve completions]]
    - [[#orderless][Orderless]]
    - [[#consult][Consult]]
    - [[#embark][Embark]]
- [[#development][Development]]
  - [[#magit][Magit]]
  - [[#projectile][Projectile]]
  - [[#treemacs][Treemacs]]
  - [[#languages][Languages]]
    - [[#clojure][Clojure]]
    - [[#javascript][Javascript]]
    - [[#lsp][LSP]]
      - [[#python][Python]]
      - [[#dap-mode][dap-mode]]
  - [[#productivity][Productivity]]
    - [[#rainbow-delimiter][Rainbow delimiter]]
    - [[#rainbow-mode][Rainbow mode]]
    - [[#flycheck][Flycheck]]
- [[#org-mode][Org mode]]
  - [[#org-config][Org config]]
  - [[#block-templates][Block templates]]
  - [[#visual-fill][Visual fill]]
  - [[#fonts-and-bullets][Fonts and bullets]]
  - [[#add-toc-automatically][Add TOC automatically]]
  - [[#pomodoro][Pomodoro]]
:END:

* Add lexical binding to generated file
#+begin_src emacs-lisp
  ;;; init.el -*- lexical-binding: t ; eval: (view-mode -1) -*-
#+end_src

* Native compilation
#+begin_src emacs-lisp
  (setq comp-async-report-warnings-errors nil)
#+end_src

* Auto update packages
#+begin_src emacs-lisp
  (use-package auto-package-update
    :config
    (setq auto-package-update-delete-old-versions t
          auto-package-update-hide-results t
          auto-package-update-prompt-before-update t)
    (auto-package-update-maybe))
#+end_src

* Keep .emacs.d clean
#+begin_src emacs-lisp
  ;; Change the user-emacs-directory to keep unwanted things out of ~/.emacs.d
  (setq user-emacs-directory (expand-file-name "~/.cache/emacs/")
      url-history-file (expand-file-name "url/history" user-emacs-directory))
  
  ;; Use no-littering to automatically set common paths to the new user-emacs-directory
  (use-package no-littering)
  
  ;; Keep customization settings in a temporary file (thanks Ambrevar!)
  (setq custom-file
      (if (boundp 'server-socket-dir)
          (expand-file-name "custom.el" server-socket-dir)
      (expand-file-name (format "emacs-custom-%s.el" (user-uid)) temporary-file-directory)))
  (load custom-file t)
#+end_src

* Server mode
#+begin_src emacs-lisp
  (server-start)
#+end_src

* Setup path from shell
#+begin_src emacs-lisp
  (use-package exec-path-from-shell
        :config
        (when (memq window-system '(mac ns x))
          (exec-path-from-shell-initialize)))
#+end_src

* Keyboard bindings
** Mac keyboard + Finnish layout 
#+begin_src emacs-lisp
  (setq mac-option-modifier nil
        mac-command-modifier 'meta
        x-select-enable-clipboard t)
#+end_src

** ESC cancels all
#+begin_src emacs-lisp
  (global-set-key (kbd "<escape>") 'keyboard-escape-quit)
#+end_src

** Evil
#+begin_src emacs-lisp
  (defun js/evil-hook ()
    (dolist (mode '(custom-mode
                    eshell-mode
                    git-rebase-mode
                    erc-mode
                    circe-server-mode
                    circe-chat-mode
                    circe-query-mode
                    sauron-mode
                    term-mode))
      (add-to-list 'evil-emacs-state-modes mode)))
  
  (defun js/dont-arrow-me-bro ()
    (interactive)
    (message "Arrow keys are bad, you know?"))
  
  (use-package undo-tree
    :init
    (global-undo-tree-mode 1))
  
  (use-package evil
    :demand t
    :init
    (setq evil-want-integration t)
    (setq evil-want-keybinding nil)
    (setq evil-want-C-u-scroll t)
    (setq evil-want-C-i-jump nil)
    (setq evil-respect-visual-line-mode t)
    (setq evil-undo-system 'undo-tree)
    :config
    (add-hook 'evil-mode-hook 'js/evil-hook)
    (evil-mode 1)
    (define-key evil-insert-state-map (kbd "C-g") 'evil-normal-state)
    (define-key evil-insert-state-map (kbd "C-h") 'evil-delete-backward-char-and-join)
  
    ;; Use visual line motions even outside of visual-line-mode buffers
    (evil-global-set-key 'motion "j" 'evil-next-visual-line)
    (evil-global-set-key 'motion "k" 'evil-previous-visual-line)
    (evil-set-initial-state 'messages-buffer-mode 'normal)
    (evil-set-initial-state 'dashboard-mode 'normal))
  
  (use-package evil-collection
    :demand t
    :after evil
    ;:init
    ;(setq evil-collection-company-use-tng nil)  ;; Is this a bug in evil-collection?
    :custom
    (evil-collection-outline-bind-tab-p nil)
    :config
    (setq evil-collection-mode-list
          (remove 'lispy evil-collection-mode-list))
    (evil-collection-init))
#+end_src

** which-key
#+begin_src emacs-lisp
  (use-package which-key
    :init (which-key-mode)
    :diminish which-key-mode
    :config
    (setq which-key-idle-delay 0.3))
#+end_src

** Leader bindings with general.el
#+begin_src emacs-lisp
  (use-package general
    :demand t
    :config
    (general-evil-setup t)
  
    (general-create-definer js/leader-key-def
      :keymaps '(normal insert visual emacs)
      :prefix "SPC"
      :global-prefix "C-SPC"))
#+end_src

** Enable keychord bind with use-package
#+begin_src emacs-lisp
  (use-package use-package-chords
    :disabled
    :config (key-chord-mode 1))
#+end_src

** Custom keybindings
#+begin_src emacs-lisp
  (js/leader-key-def
        "f"   '(:ignore t :which-key "files")
        "ff"  '(find-file :which-key "open file")
        "fs"  'save-buffer
        "fr"  '(consult-recent-file :which-key "recent files")
        "fR"  '(revert-buffer :which-key "revert file"))
#+end_src

* General configuration
** Window size
#+begin_src emacs-lisp
  (when window-system (set-frame-size (selected-frame) 129 80))
  (js/leader-key-def
    "w"  '(:ignore t :which-key "window")
    "wm" '(toggle-frame-maximized :which-key "maxized")
    "wf" '(toggle-frame-fullscreen :which-key "fullscreen"))
#+end_src
** Dashboard
#+begin_src emacs-lisp
  (use-package dashboard
    :demand t
    :preface
    (setq js/startup-time-message
          (let ((package-count (hash-table-size straight--profile-cache)))
            (format "Emacs loaded in %s with %d packages and %d garbage collections."
                    (format "%.2f seconds"
                            (float-time
                             (time-subtract after-init-time before-init-time)))
                    package-count
                    gcs-done)))
    :config
    (setq dashboard-startup-banner 'logo
          dashboard-projects-backend 'projectile
          dashboard-set-heading-icons t
          dashboard-set-file-icons t
          dashboard-center-content t
          dashboard-items '((recents  . 5)
                            (bookmarks . 5)
                            (projects . 5)
                            (agenda . 5)
                            (registers . 5))
          dashboard-init-info js/startup-time-message)
    (dashboard-setup-startup-hook)
    (setq initial-buffer-choice (lambda () (get-buffer-create "*dashboard*"))))
#+end_src

** Improve scrolling
#+begin_src emacs-lisp
  (setq mouse-wheel-scroll-amount '(1 ((shift) . 1))) ;; one line at a time
  (setq mouse-wheel-progressive-speed nil) ;; don't accelerate scrolling
  (setq mouse-wheel-follow-mouse 't) ;; scroll window under mouse
  (setq scroll-step 1) ;; keyboard scroll one line at a time
  (setq use-dialog-box nil)
#+end_src

** Line numbers
#+begin_src emacs-lisp
  (column-number-mode)
  
  ;; Enable line numbers for some modes
  (dolist (mode '(text-mode-hook
                  prog-mode-hook
                  conf-mode-hook))
    (add-hook mode (lambda () (display-line-numbers-mode 1))))
  
  ;; Override some modes which derive from the above
  (dolist (mode '(org-mode-hook))
    (add-hook mode (lambda () (display-line-numbers-mode 0))))
  #+end_src
  
** Theme
#+begin_src emacs-lisp
  (use-package spacegray-theme)
  (use-package doom-themes)
  (load-theme 'doom-one t)
  (doom-themes-visual-bell-config)
#+end_src

** Font
#+begin_src emacs-lisp
  (defvar efs/default-font-size 150)
  (defvar efs/default-variable-font-size 150)
  
  (set-face-attribute 'default nil
                      :font "JetBrains Mono"
                      :weight 'light
                      :height efs/default-font-size)
  
  ;; Set the fixed pitch face
  (set-face-attribute 'fixed-pitch nil
                      :font "JetBrains Mono"
                      :weight 'light
                      :height efs/default-font-size)
  
  ;; Set the variable pitch face
  (set-face-attribute 'variable-pitch nil
                      :font "Iosevka Aile"
                      :height efs/default-variable-font-size
                      :weight 'light)
#+end_src

** Emojis in buffers
#+begin_src emacs-lisp
  (use-package emojify
    :hook (erc-mode . emojify-mode)
    :commands emojify-mode)
#+end_src

** Doom modeline
Run =all-the-icons-install-fonts= after installing this package
#+begin_src emacs-lisp
  (use-package minions
    :hook (doom-modeline-mode . minions-mode))
  
  (use-package doom-modeline
    :after eshell     ;; Make sure it gets hooked after eshell
    :hook (after-init . doom-modeline-mode)
    :custom
    (doom-modeline-lsp t)
    (doom-modeline-github nil)
    (doom-modeline-mu4e nil)
    (doom-modeline-irc nil)
    (doom-modeline-minor-modes t)
    (doom-modeline-persp-name nil)
    (doom-modeline-buffer-file-name-style 'truncate-except-project)
    (doom-modeline-major-mode-icon nil))
  (doom-modeline-mode 1)
#+end_src

** Mode diminishing
#+begin_src emacs-lisp
  (use-package diminish)
#+end_src

** Recent files
#+begin_src emacs-lisp
  (recentf-mode 1)
  (setq recentf-max-menu-items 25)
  (setq recentf-max-saved-items 25)
#+end_src

** Restarting and quitting Emacs
#+begin_src emacs-lisp
  (defun js/reload-init ()
    "Reload init.el."
    (interactive)
    (message "Reloading init.el...")
    (load user-init-file nil 'nomessage)
    (message "Reloading init.el... done."))

  (use-package restart-emacs
    :general
    (js/leader-key-def
      "q"   '(:ignore t :which-key "quit")
      "qq"  '(save-buffers-kill-emacs :which-key "quit emacs")
      "qR"  'restart-emacs
      "qr"  '(js/reload-init :which-key "reload confs")))
#+end_src

* Completion
Many of the settings here are taken from daviwil.
https://github.com/daviwil/dotfiles/blob/master/Emacs.org#completion-system

** Vertico
#+begin_src emacs-lisp
  (defun js/minibuffer-backward-kill (arg)
    "When minibuffer is completing a file name delete up to parent
  folder, otherwise delete a word"
    (interactive "p")
    (if minibuffer-completing-file-name
        ;; Borrowed from https://github.com/raxod502/selectrum/issues/498#issuecomment-803283608
        (if (string-match-p "/." (minibuffer-contents))
            (zap-up-to-char (- arg) ?/)
          (delete-minibuffer-contents))
        (backward-kill-word arg)))
  
  (use-package vertico
    :bind (:map vertico-map
           ("C-j" . vertico-next)
           ("C-k" . vertico-previous)
           ("C-f" . vertico-exit)
           :map minibuffer-local-map
           ("M-h" . js/minibuffer-backward-kill))
    :custom
    (vertico-cycle t)
    :custom-face
    (vertico-current ((t (:background "#3a3f5a"))))
    :init
    (vertico-mode))
#+end_src

** Corfu
#+begin_src emacs-lisp
  (use-package corfu
    :straight '(corfu :host github
                      :repo "minad/corfu")
    :bind (:map corfu-map
           ("C-j" . corfu-next)
           ("C-k" . corfu-previous)
           ("C-f" . corfu-insert))
    :custom
    (corfu-cycle t)
    :config
    (corfu-global-mode))
#+end_src

** Save minibuffer history
 #+begin_src emacs-lisp 
   (use-package savehist
     :demand 
     :config
     (setq history-length 25)
     (savehist-mode 1))
   
     ;; Individual history elements can be configured separately
     ;;(put 'minibuffer-history 'history-length 25)
     ;;(put 'evil-ex-history 'history-length 50)
     ;;(put 'kill-ring 'history-length 25))
 #+end_src

** Completion metadata with Marginalia
#+begin_src emacs-lisp 
  (use-package marginalia
    :after vertico
    :custom
    (marginalia-annotators '(marginalia-annotators-heavy marginalia-annotators-light nil))
    :init
    (marginalia-mode))
#+end_src

** Improve completions
*** Orderless
"This package provides an orderless completion style that divides the pattern into space-separated components, and matches candidates that match all of the components in any order."
https://github.com/oantolin/orderless
#+begin_src emacs-lisp
  (use-package orderless
    :demand t
    :init
    (setq completion-styles '(orderless)
          completion-category-defaults nil
          completion-category-overrides '((file (styles . (partial-completion))))))
#+end_src

*** Consult
"Consult provides various practical commands based on the Emacs completion function completing-read, which allows to quickly select an item from a list of candidates with completion."
https://github.com/minad/consult
#+begin_src emacs-lisp
  (defun js/get-project-root ()
    (when (fboundp 'projectile-project-root)
      (projectile-project-root)))
  
  (use-package consult
    :bind (("C-s" . consult-line)
           ("C-M-l" . consult-imenu)
           :map minibuffer-local-map
           ("C-r" . consult-history))
    :custom
    (consult-project-root-function #'js/get-project-root)
    (completion-in-region-function #'consult-completion-in-region))
#+end_src

*** Embark

config.el
(require 'package)
(setq package-archives '(("melpa" . "https://melpa.org/packages/")
                         ("melpa-stable" . "https://stable.melpa.org/packages/")
                         ("org" . "https://orgmode.org/elpa/")
                         ("elpa" . "https://elpa.gnu.org/packages/")))
(package-initialize)
;; (unless package-archive-contents
;;   (package-refresh-contents))

(unless (package-installed-p 'use-package)
  (package-install 'use-package))
(require 'use-package)
(setq use-package-always-ensure t)

(defvar bootstrap-version)
(let ((bootstrap-file
      (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
      (bootstrap-version 5))
  (unless (file-exists-p bootstrap-file)
    (with-current-buffer
        (url-retrieve-synchronously
        "https://raw.githubusercontent.com/raxod502/straight.el/develop/install.el"
        'silent 'inhibit-cookies)
      (goto-char (point-max))
      (eval-print-last-sexp)))
  (load bootstrap-file nil 'nomessage))
;; Always use straight to install on systems other than Linux
(setq straight-use-package-by-default (not (eq system-type 'gnu/linux)))

;; Use straight.el for use-package expressions
(straight-use-package 'use-package)

;; Load the helper package for commands like `straight-x-clean-unused-repos'
(require 'straight-x)

;; (setq package-native-compile t)

(use-package doom-themes :defer t)
(setq doom-themes-enable-bold t    ; if nil, bold is universally disabled
      doom-themes-enable-italic t) ; if nil, italics is universally disabled
(load-theme 'doom-one t)

(defun my/apply-theme (appearance)
 "Load theme, taking current system APPEARANCE into consideration."
  (mapc #'disable-theme custom-enabled-themes)
 (pcase appearance
   ('light (load-theme 'tango-dark t)) ;; or 'tango
   ('dark (load-theme 'tango-dark t))))

(add-hook 'ns-system-appearance-change-functions #'my/apply-theme)
(add-to-list 'default-frame-alist '(ns-transparent-titlebar . t))
(add-to-list 'default-frame-alist '(ns-appearance . dark))

;;(add-to-list 'default-frame-alist '(undecorated . 1))
(setq ns-use-proxy-icon nil)
(setq frame-title-format nil)

(setq comp-async-report-warnings-errors nil)

(defvar efs/default-font-size 150)
(defvar efs/default-variable-font-size 150)

(set-face-attribute 'default nil :font "JetBrains Mono" :height efs/default-font-size)

;; Set the fixed pitch face
(set-face-attribute 'fixed-pitch nil :font "JetBrains Mono" :height efs/default-font-size)

;; Set the variable pitch face
(set-face-attribute 'variable-pitch nil :font "JetBrains Mono" :height efs/default-font-size :weight 'regular)

(use-package all-the-icons)

(use-package dashboard
 :init      ;; tweak dashboard config before loading it
 (add-hook 'after-init-hook 'dashboard-refresh-buffer)
 (setq dashboard-set-heading-icons t)
 (setq dashboard-set-file-icons t)
 ;; (setq dashboard-banner-logo-title "Emacs Is More Than A Text Editor!")
 ;; (setq dashboard-startup-banner 'logo) ;; use standard emacs logo as banner
 ;; (setq dashboard-startup-banner "~/.emacs.d/emacs-dash.png")  ;; use custom image as banner
 ;; (setq dashboard-banner-logo-title "Welcome")
 (setq dashboard-startup-banner 'logo)
 (setq dashboard-center-content nil)
 (setq dashboard-items '((recents . 5)
                         (agenda . 5 )
                         (bookmarks . 3)
                         (projects . 3)
                         (registers . 3)))
 :config
 (dashboard-setup-startup-hook)
 (dashboard-modify-heading-icons '((recents . "file-text")
                             (bookmarks . "book"))))
;; (setq initial-buffer-choice (lambda () (get-buffer "*dashboard*")))

;; (setq scroll-conservatively 101) ;; value greater than 100 gets rid of half page jumping
;; (setq mouse-wheel-scroll-amount '(3 ((shift) . 6))) ;; how many lines at a time
;; (setq mouse-wheel-progressive-speed nil) ;; accelerate scrolling
;; (setq mouse-wheel-follow-mouse 't) ;; scroll window under mouse
;; (setq inhibit-compacting-font-caches t)

;; (use-package ivy
  ;;   :diminish
  ;;   :bind (("C-s" . swiper)
  ;;          :map ivy-minibuffer-map
  ;;          ("TAB" . ivy-alt-done)
  ;;          ("C-l" . ivy-alt-done)
  ;;          ("C-j" . ivy-next-line)
  ;;          ("C-k" . ivy-previous-line)
  ;;          :map ivy-switch-buffer-map
  ;;          ("C-k" . ivy-previous-line)
  ;;          ("C-l" . ivy-done)
  ;;          ("C-d" . ivy-switch-buffer-kill)
  ;;          :map ivy-reverse-i-search-map
  ;;          ("C-k" . ivy-previous-line)
  ;;          ("C-d" . ivy-reverse-i-search-kill))
  ;;   :config
  ;;   (ivy-mode 1))

  ;; (use-package ivy-rich
  ;;   :after ivy
  ;;   :init
  ;;   (ivy-rich-mode 1))

  ;; (use-package counsel
  ;;   :bind (("C-M-j" . 'counsel-switch-buffer)
  ;;          :map minibuffer-local-map
  ;;          ("C-r" . 'counsel-minibuffer-history))
  ;;   :custom
  ;;   (counsel-linux-app-format-function #'counsel-linux-app-format-function-name-only)
  ;;   :config
  ;;   (counsel-mode 1))
  (use-package savehist
  :config
  (setq history-length 25)
  (savehist-mode 1))

  ;; Individual history elements can be configured separately
  ;;(put 'minibuffer-history 'history-length 25)
  ;;(put 'evil-ex-history 'history-length 50)
  ;;(put 'kill-ring 'history-length 25))

  (defun js/minibuffer-backward-kill (arg)
  "When miibuffer is completing a file name delete up to parent
folder, otherwise delete a word"
  (interactive "p")
  (if minibuffer-completing-file-name
      ;; Borrowed from https://github.com/raxod502/selectrum/issues/498#issuecomment-803283608
      (if (string-match-p "/." (minibuffer-contents))
          (zap-up-to-char (- arg) ?/)
        (delete-minibuffer-contents))
      (backward-kill-word arg)))

(use-package vertico
  :straight '(vertico :host github
                      :repo "minad/vertico"
                      :branch "main")
  :bind (:map vertico-map
         ("C-j" . vertico-next)
         ("C-k" . vertico-previous)
         ("C-f" . vertico-exit)
         :map minibuffer-local-map
         ("M-h" . dw/minibuffer-backward-kill))
  :custom
  (vertico-cycle t)
  :custom-face
  (vertico-current ((t (:background "#3a3f5a"))))
  :init
  (vertico-mode))

(use-package corfu
  :straight '(corfu :host github
                    :repo "minad/corfu")
  :bind (:map corfu-map
         ("C-j" . corfu-next)
         ("C-k" . corfu-previous)
         ("C-f" . corfu-insert))
  :custom
  (corfu-cycle t)
  :config
  (corfu-global-mode))

(use-package orderless
:straight t
:init
(setq completion-styles '(orderless)
      completion-category-defaults nil
      completion-category-overrides '((file (styles . (partial-completion))))))

;; (setq ivy-initial-inputs-alist nil)

;; (use-package ivy-posframe
;;   :init
;;   (setq ivy-posframe-display-functions-alist
;;     '((swiper                     . ivy-posframe-display-at-point)
;;       (complete-symbol            . ivy-posframe-display-at-point)
;;       (counsel-M-x                . ivy-display-function-fallback)
;;       (counsel-esh-history        . ivy-posframe-display-at-window-center)
;;       (counsel-describe-function  . ivy-display-function-fallback)
;;       (counsel-describe-variable  . ivy-display-function-fallback)
;;       (counsel-find-file          . ivy-display-function-fallback)
;;       (counsel-recentf            . ivy-display-function-fallback)
;;       (counsel-register           . ivy-posframe-display-at-frame-bottom-window-center)
;;       (dmenu                      . ivy-posframe-display-at-frame-top-center)
;;       (nil                        . ivy-posframe-display))
;;     ivy-posframe-height-alist
;;     '((swiper . 20)
;;       (dmenu . 20)
;;       (t . 10)))
;;   :config
;;   (ivy-posframe-mode 1)) ; 1 enables posframe-mode, 0 disables it.

;; (use-package ivy-prescient
;;   :after counsel
;;   :custom
;;   (ivy-prescient-enable-filtering nil)
;;   :config
;;   ;; Uncomment the following line to have sorting remembered across sessions!
;;   ;(prescient-persist-mode 1)
;;   (ivy-prescient-mode 1))

(use-package which-key
  :diminish which-key-mode
  :config
  (setq which-key-idle-delay 0.3))

(use-package projectile
  :diminish projectile-mode
  :config (projectile-mode)
  :custom ((projectile-completion-system 'ivy))
  :bind-keymap
  ("C-c p" . projectile-command-map)
  :init
  ;; NOTE: Set this to the folder where you keep your Git repos!
  (when (file-directory-p "~/Dev")
    (setq projectile-project-search-path '("~/Dev")))
  (setq projectile-switch-project-action #'projectile-dired))

(use-package counsel-projectile
  :config (counsel-projectile-mode))

(use-package evil-commentary)
(evil-commentary-mode)

;; (defun efs/org-font-setup ()
;;   ;; Replace list hyphen with dot
;;   (font-lock-add-keywords 'org-mode
;;                           '(("^ *\\([-]\\) "
;;                              (0 (prog1 () (compose-region (match-beginning 1) (match-end 1) "•"))))))

;;   ;; Set faces for heading levels
;;   (dolist (face '((org-level-1 . 1.2)
;;                   (org-level-2 . 1.1)
;;                   (org-level-3 . 1.05)
;;                   (org-level-4 . 1.0)
;;                   (org-level-5 . 1.1)
;;                   (org-level-6 . 1.1)
;;                   (org-level-7 . 1.1)
;;                   (org-level-8 . 1.1)))
;;     (set-face-attribute (car face) nil :font "JetBrains Mono" :weight 'regular :height (cdr face)))

;;   ;; Ensure that anything that should be fixed-pitch in Org files appears that way
;;   (set-face-attribute 'org-block nil    :foreground nil :inherit 'fixed-pitch)
;;   (set-face-attribute 'org-table nil    :inherit 'fixed-pitch)
;;   (set-face-attribute 'org-formula nil  :inherit 'fixed-pitch)
;;   (set-face-attribute 'org-code nil     :inherit '(shadow fixed-pitch))
;;   (set-face-attribute 'org-table nil    :inherit '(shadow fixed-pitch))
;;   (set-face-attribute 'org-verbatim nil :inherit '(shadow fixed-pitch))
;;   (set-face-attribute 'org-special-keyword nil :inherit '(font-lock-comment-face fixed-pitch))
;;   (set-face-attribute 'org-meta-line nil :inherit '(font-lock-comment-face fixed-pitch))
;;   (set-face-attribute 'org-checkbox nil  :inherit 'fixed-pitch)
;;   (set-face-attribute 'line-number nil :inherit 'fixed-pitch)
;;   (set-face-attribute 'line-number-current-line nil :inherit 'fixed-pitch))

;; (defun efs/org-mode-setup ()
;;   (org-indent-mode)
;;   (variable-pitch-mode 1)
;;   (visual-line-mode 1))

;; (use-package org
;;   :hook (org-mode . efs/org-mode-setup)
;;   :config
;;   (setq org-ellipsis " ▾")

;;   (setq org-agenda-start-with-log-mode t)
;;   (setq org-log-done 'time)
;;   (setq org-log-into-drawer t)

;;   (require 'org-habit)
;;   (add-to-list 'org-modules 'org-habit)
;;   (setq org-habit-graph-column 60)

;;   (setq org-todo-keywords
;;     '((sequence "TODO(t)" "NEXT(n)" "|" "DONE(d!)")
;;       (sequence "BACKLOG(b)" "PLAN(p)" "READY(r)" "ACTIVE(a)" "REVIEW(v)" "WAIT(w@/!)" "HOLD(h)" "|" "COMPLETED(c)" "CANC(k@)")))

;;   (setq org-refile-targets
;;     '(("Archive.org" :maxlevel . 1)
;;       ("Tasks.org" :maxlevel . 1)))

;;   ;; Save Org buffers after refiling!
;;   (advice-add 'org-refile :after 'org-save-all-org-buffers)

;;   (setq org-tag-alist
;;     '((:startgroup)
;;        ; Put mutually exclusive tags here
;;        (:endgroup)
;;        ("@errand" . ?E)
;;        ("@home" . ?H)
;;        ("@work" . ?W)
;;        ("agenda" . ?a)
;;        ("planning" . ?p)
;;        ("publish" . ?P)
;;        ("batch" . ?b)
;;        ("note" . ?n)
;;        ("idea" . ?i)))

;;   ;; Configure custom agenda views
;;   (setq org-agenda-custom-commands
;;    '(("d" "Dashboard"
;;      ((agenda "" ((org-deadline-warning-days 7)))
;;       (todo "NEXT"
;;         ((org-agenda-overriding-header "Next Tasks")))
;;       (tags-todo "agenda/ACTIVE" ((org-agenda-overriding-header "Active Projects")))))

;;     ("n" "Next Tasks"
;;      ((todo "NEXT"
;;         ((org-agenda-overriding-header "Next Tasks")))))

;;     ("W" "Work Tasks" tags-todo "+work-email")

;;     ;; Low-effort next actions
;;     ("e" tags-todo "+TODO=\"NEXT\"+Effort<15&+Effort>0"
;;      ((org-agenda-overriding-header "Low Effort Tasks")
;;       (org-agenda-max-todos 20)
;;       (org-agenda-files org-agenda-files)))

;;     ("w" "Workflow Status"
;;      ((todo "WAIT"
;;             ((org-agenda-overriding-header "Waiting on External")
;;              (org-agenda-files org-agenda-files)))
;;       (todo "REVIEW"
;;             ((org-agenda-overriding-header "In Review")
;;              (org-agenda-files org-agenda-files)))
;;       (todo "PLAN"
;;             ((org-agenda-overriding-header "In Planning")
;;              (org-agenda-todo-list-sublevels nil)
;;              (org-agenda-files org-agenda-files)))
;;       (todo "BACKLOG"
;;             ((org-agenda-overriding-header "Project Backlog")
;;              (org-agenda-todo-list-sublevels nil)
;;              (org-agenda-files org-agenda-files)))
;;       (todo "READY"
;;             ((org-agenda-overriding-header "Ready for Work")
;;              (org-agenda-files org-agenda-files)))
;;       (todo "ACTIVE"
;;             ((org-agenda-overriding-header "Active Projects")
;;              (org-agenda-files org-agenda-files)))
;;       (todo "COMPLETED"
;;             ((org-agenda-overriding-header "Completed Projects")
;;              (org-agenda-files org-agenda-files)))
;;       (todo "CANC"
;;             ((org-agenda-overriding-header "Cancelled Projects")
;;              (org-agenda-files org-agenda-files)))))))

;; (define-key global-map (kbd "C-c j")
;;   (lambda () (interactive) (org-capture nil "jj")))

;; (efs/org-font-setup))

;; (use-package org-tempo
;;   :ensure nil)

;; (use-package org-bullets
;;   :after or
;;   :hook (org-mode . org-bullets-mode)
;;   :custom
;;   (org-bullets-bullet-list '("◉" "○" "●" "○" "●" "○" "●")))

(org-babel-do-load-languages
  'org-babel-load-languages
  '((emacs-lisp . t)))

(push '("conf-unix" . conf-unix) org-src-lang-modes)

;; Syntax highlight in #+BEGIN_SRC blocks
(setq org-src-fontify-natively t)
;; Don't prompt before running code in org
(setq org-confirm-babel-evaluate nil)

;; Make ESC quit prompts
(global-set-key (kbd "<escape>") 'keyboard-escape-quit)

(use-package general
  :config
  (general-create-definer rune/leader-keys
    :keymaps '(normal insert visual emacs)
    :prefix "SPC"
    :global-prefix "C-SPC")

  (rune/leader-keys
    "t"  '(:ignore t :which-key "toggles")
    "tt" '(counsel-load-theme :which-key "choose theme")))

(use-package evil
  :init
  (setq evil-want-integration t)
  (setq evil-want-keybinding nil)
  (setq evil-vsplit-window-right t)
  (setq evil-split-window-below t)
  (setq evil-want-C-u-scroll t)
  (setq evil-want-C-i-jump nil)
  :config
  (evil-mode 1)
  (define-key evil-insert-state-map (kbd "C-g") 'evil-normal-state)
  (define-key evil-insert-state-map (kbd "C-h") 'evil-delete-backward-char-and-join)

;; Use visual line motions even outside of visual-line-mode buffers
(evil-global-set-key 'motion "j" 'evil-next-visual-line)
(evil-global-set-key 'motion "k" 'evil-previous-visual-line)

(evil-set-initial-state 'messages-buffer-mode 'normal)
(evil-set-initial-state 'dashboard-mode 'normal))

(use-package evil-collection
  :after evil
  :config
  (evil-collection-init))

(use-package evil-tutor)

(use-package general
  :config
  (general-evil-setup t))

(nvmap :keymaps 'override :prefix "SPC"
       "SPC"   '(counsel-M-x :which-key "M-x")
       "c c"   '(compile :which-key "Compile")
       "c C"   '(recompile :which-key "Recompile")
       "h r r" '((lambda () (interactive) (load-file "~/.emacs.default/init.el")) :which-key "Reload emacs config")
       "t t"   '(toggle-truncate-lines :which-key "Toggle truncate lines"))
(nvmap :keymaps 'override :prefix "SPC"
       "m *"   '(org-ctrl-c-star :which-key "Org-ctrl-c-star")
       "m +"   '(org-ctrl-c-minus :which-key "Org-ctrl-c-minus")
       "m ."   '(counsel-org-goto :which-key "Counsel org goto")
       "m e"   '(org-export-dispatch :which-key "Org export dispatch")
       "m f"   '(org-footnote-new :which-key "Org footnote new")
       "m h"   '(org-toggle-heading :which-key "Org toggle heading")
       "m i"   '(org-toggle-item :which-key "Org toggle item")
       "m n"   '(org-store-link :which-key "Org store link")
       "m o"   '(org-set-property :which-key "Org set property")
       "m t"   '(org-todo :which-key "Org todo")
       "m x"   '(org-toggle-checkbox :which-key "Org toggle checkbox")
       "m B"   '(org-babel-tangle :which-key "Org babel tangle")
       "m I"   '(org-toggle-inline-images :which-key "Org toggle inline imager")
       "m T"   '(org-todo-list :which-key "Org todo list")
       "o a"   '(org-agenda :which-key "Org agenda")
       )

(setq mac-option-modifier nil
      mac-command-modifier 'meta
      x-select-enable-clipboard t)

(scroll-bar-mode -1)
(tool-bar-mode -1)
(tooltip-mode -1)
(set-fringe-mode 0)
(menu-bar-mode 1)
(column-number-mode)
(global-display-line-numbers-mode t)

;; Disable line numbers for some modes
(dolist (mode '(org-mode-hook
                term-mode-hook
                shell-mode-hook
                eshell-mode-hook))
 (add-hook mode (lambda () (display-line-numbers-mode 0))))
